src/libraries/Microsoft.Bcl.TimeProvider/src/System/Threading/Tasks/TimeProviderTaskExtensions.cs

```diff
@@ -89,7 +89,14 @@ public static Task Delay(this TimeProvider timeProvider, TimeSpan delay, Cancell
             state.Registration = cancellationToken.Register(static delayState =>
             {
                 DelayState s = (DelayState)delayState!;
-                s.TrySetCanceled(s.CancellationToken);
+
+                // When cancellation is requested, we need to force the task continuation to run asynchronously
```

Comment:
```suggestion
                // When cancellation is requested, we need to force the task continuation to run asynchronously
                // to avoid doing arbitrary amounts of work as part of a call to CancellationTokenSource.Cancel.
```